cli/functionary/parser/python.py

```diff
@@ -0,0 +1,191 @@
+    except FileNotFoundError:
+        raise click.ClickException("Could not find functions.py")
+    except PermissionError:
+        raise click.ClickException("Did not have permission to access functions.py")
+    return function_data
+
+
+def _retrieve_function_nodes(data):
+    """
+    Helper function for py_parse to retrieve
+    ast function nodes.
+
+    Args:
+        data: string representing python code
+    Returns:
+        function_nodes: list containing top level
+        ast function nodes
+    Raises:
+        ClickException if incorrect python syntax in data
+    """
+    function_nodes = []
+    try:
+        for node in ast.parse(data).body:
+            if isinstance(node, ast.FunctionDef):
+                function_nodes.append(node)
+    except SyntaxError:
+        raise click.ClickException(
+            "Syntax error in function.py file. Please correct code and try again"
+        )
+    return function_nodes
+
+
+def _create_argument_list(node):
+    """
+    Helper function for py_parse to create
+    a list of dictionaries representing function
+    args. If parser cannot detect arg type, alert
+    user that they must manually fill out package.yaml
+
+    Args:
+        node: ast node representing function
+    Returns:
+        arg_list: list of dictionaries containing arg info
+    """
+    args = node.args.args
+    arg_list = []
+
+    for arg in args:
+        arg_dict = {}
+        arg_dict["name"] = arg.arg
+
+        type = None
+        if arg.annotation:
+            try:
+                if isinstance(arg.annotation, ast.Name):
+                    type = arg.annotation.id
+                elif isinstance(arg.annotation, ast.Attribute):
+                    type = arg.annotation.attr
+
+                if type:
+                    arg_dict["type"] = type_map[type]
+            except KeyError:
+                type = None
+
+        arg_dict["required"] = True
+        arg_list.append(arg_dict)
+
+        if type is None:
+            click.echo(
+                f"Type of argument {arg.arg} in function {node.name} not detected or "
+                "is unsupported. Please specify type in package.yaml manually before "
+                "publishing package, as argument type is required.\n"
+            )
+    return arg_list
+
+
+def _assign_defaults(node, arg_list):
+    """
+    Helper function for py_parse. If arg
+    has a default, fill out arg's dictionary with
+    default
+
+    Args:
+        node: ast node representing function
+        arg_list: list of dictionaries representing args
+    Returns:
+        arg_list: list of dictionaries representing args,
+        now with defaults filled out as needed
+    """
+    defaults = node.args.defaults
+
+    default_start = len(arg_list) - len(defaults)
+    has_default = arg_list[default_start:]
+    no_default = arg_list[:default_start]
+
+    for arg, default in zip(has_default, defaults):
+        if arg["type"] == type_map["date"]:
```

Comment:
If `arg["type"]` doesn't exist just use default.value, otherwise this line explodes.